src/core/regimes.rkt

```diff
@@ -56,13 +56,13 @@
   (debug "Found split indices:" best #:from 'regime #:depth 3)
   best)
 
-(define (exprs-to-branch-on alts)
+(define (exprs-to-branch-on alts repr)
   (define alt-critexprs (map (compose all-critical-subexpressions alt-program) alts))
   (define start-critexprs (all-critical-subexpressions (*start-prog*)))
   ;; We can only binary search if the branch expression is critical
   ;; for all of the alts and also for the start prgoram.
   (filter
-   (λ (e) (equal? (type-name (representation-type (get-representation* (type-of e (*var-reprs*))))) 'real))
+   (λ (e) (set-member? '(binary64 binary32) (type-of e repr (*var-reprs*))))
    (set-intersect start-critexprs (apply set-union alt-critexprs))))
```

Comment:
This change isn't correct. We're interested not in the representation of `e` but in its type: it needs to be real, because complex numbers and booleans aren't sorted. I think we need a `type-of` for type and a `repr-of` for representations.